Enable dynamic linking behind a feature flag

tronje opened the pull request:
I needed to be able to link to a system libsrtp2 for a project, so I added a Cargo feature that enables dynamic linking for this crate.

tronje committed:
Enable dynamic linking behind a feature flag

Signed-off-by: Tronje Krabbe <tkrabbe@jusst.de>

HyeonuPark commented:
Thanks for your contribution! I'm not a `-sys` crate expert so I just want to double check here.

1. I see `cargo:rustc-link-lib=srtp2` in it, but it seems you want to enable dynamic linking only. Shouldn't it be `=dylib=srtp2`?

1. `pkg-config` integration seems mandatory for linking with system libs on unix-likes. Will it works out of the box on most systems without pkg-config?

1. Cargo features can't be easily configured on end binary crate. Maybe it's better to define some environment variable named like `RUST_SRTP2_SYS_USE_SYSTEM_LIB` to enable this feature?

I'm open to your suggestions here. Thanks!

tronje commented:
Hi! Thanks for the response. I'm not an expert either, but let's see.

1. I assumed that the default was dynamic linking, but I was wrong. It should indeed by `[...]=dylib=srtp2`. I'll fix that.

2. It seems to work fine for me, tested both on a "normal" Linux system, and in a Buildroot environment. I agree, though, that it may be better and more robust to use `pkg-config` to search for the search path and set it explicitly. I'll get to work on that.

3. Yes, that is a bit tricky. I don't love the idea of using a magic environment variable when Cargo already has the `features` feature, but then you are of course completely correct that it is impossible to set the feature as a consumer of just the `srtp` crate... Your call!

tronje committed:
Use pkg-config to generate rustc flags when linking dynamically

Signed-off-by: Tronje Krabbe <tkrabbe@jusst.de>

tronje committed:
Use system includes to generate bindings when linking dynamically

This also adds a wrapper header that includes the expected headers for
us, as is demonstrated in the bindgen tutorial.

Signed-off-by: Tronje Krabbe <tkrabbe@jusst.de>

tronje commented:
I ran into some tricky issues while building the `pkg-config` integration.
So for one thing we would want to generate the bindings from the headers in `/usr/include/srtp` when linking dynamically, to make sure that the bindings are consistent with the library that we link with. However, that introduces an issue where we then cannot include `srtp_priv.h` in the generated bindings, leading to a mismatch with the bindings generated by the static version from the submodule.

I've pushed my changes so you can have a look. But to be honest, this is diverging pretty far from your original crate, to the point where I'm not sure if it wouldn't be better for me to just create my own crate and not pollute yours...

tronje committed:
Add a feature that allows using pre-generated bindings

The bindings are checked in, and were generated from libsrtp 2.3.0.
The feature skips any work the build.rs would have done, and instead
simply exposes the pre-generated bindings completely.

Signed-off-by: Tronje Krabbe <tkrabbe@jusst.de>

tronje committed:
rustfmt build.rs

Signed-off-by: Tronje Krabbe <tkrabbe@jusst.de>

tronje committed:
Fix build.rs for pre-generated-bindings feature

Even though the bindings are pre-generated, we must still link against
libsrtp, which was omitted mistakenly.

Signed-off-by: Tronje Krabbe <tkrabbe@jusst.de>
